File: Vector.cpp
#include<math.h>
#include "Constants.cpp"
class Vector {
    public:
        Vector();
        Vector(double x, double y);
        Vector(double x, double y, double z);
        double x, y, z;
        Vector rotateTheta(double theta);
        Vector operator + (Vector v);
        void operator += (Vector v);
        Vector operator - (Vector v);
        void operator -= (Vector v);
        Vector operator * (double q);
        void operator *= (double q);
        Vector operator / (double q);
        double abs();
        Vector normalise();
        void print();
        void print(char* end);
        
};
Vector::Vector(){

}
Vector::Vector(double x_, double y_) {
    this->x = x_; this->y=y_; this->z=0;
}

Vector::Vector(double x_, double y_, double z_) {
    this->x = x_; this->y=y_; this->z=z_;
}

Vector Vector::rotateTheta(double theta) {
    // theta radians
    double x_ = this->x*cos(theta) - this->y*sin(theta);
    double y_ = this->x*sin(theta) + this->y*cos(theta);
    return Vector(x_, y_, this->z);
}

Vector Vector::operator +(Vector v) {
    return Vector(this->x+v.x, this->y+v.y, this->z+v.z);
}
void Vector::operator +=(Vector v) {
    this->x+=v.x;
    this->y+=v.y;
    this->z+=v.z;
}
Vector Vector::operator -(Vector v) {
    return Vector(this->x-v.x, this->y-v.y, this->z-v.z);
}
void Vector::operator -=(Vector v) {
    this->x-=v.x;
    this->y-=v.y;
    this->z-=v.z;
}
Vector Vector::operator *(double q) {
    return Vector(this->x*q, this->y*q, this->z*q);
}
void Vector::operator *= (double q){
    this->x *= q;
    this->y *= q;
    this->z *= q;
}
Vector Vector::operator /(double q) {
    return Vector(this->x/q, this->y/q, this->z/q);
}
double Vector::abs() {
    return pow(pow(this->x,2)+pow(this->y, 2)+pow(this->z,2), 0.5);
}
Vector Vector::normalise() {
    return *this / this->abs();
}
void Vector::print(){
    std::cout << "(" << this->x << ", " << this->y << ", " << this->z << ")" << std::endl;
}
void Vector::print(char* end){
    std::cout << "(" << this->x << ", " << this->y << ", " << this->z << ")" << end;
}

Vector VectorFromSpherical(double r, double theta, double phi) {
    double x = r*cos(theta)*sin(phi);
    double y = r*sin(theta)*sin(phi);
    double z = r*cos(phi);
    return Vector(x, y, z);
}

File: Vector.py
from numpy import degrees, radians, cos, sin, arccos, arctan2

class Vector():
    def __init__(self, x, y, z=0):
        '''inherently 3d vector, z can be left off to make a 2d vector'''
        self.x = x
        self.y = y
        self.z = z

    def rotateTheta(self, theta):
        theta = radians(theta)
        x = self.x*cos(theta) - self.y*sin(theta)
        y = self.x*sin(theta) + self.y*cos(theta)
        return Vector(x, y)

    def rotatePhi(self, phi):
        phi = radians(phi)
        r = (self.x**2+self.y**2+self.z**2)**0.5
        theta = arccos(self.z / r)
        phi = arctan2(self.y, self.x) + phi
        theta = degrees(theta)
        phi = degrees(phi)
        return Vector.fromspherical(r, theta, phi)

    @classmethod
    def fromspherical(cls, r, theta, phi):
        theta = radians(theta)
        phi = radians(phi)
        return r * cls(cos(theta)*sin(phi),
            sin(theta)*sin(phi),
            cos(phi))

    def __str__(self):
        return f"({self.x}, {self.y}, {self.z})"
    __repr__=__str__

    def __add__(self, other):
        if isinstance(other, Vector):
            x = self.x + other.x
            y = self.y + other.y
            z = self.z + other.z
            return Vector(x, y, z)
        else:
            return NotImplemented

    def __sub__(self, other):
        if isinstance(other, Vector):
            x = self.x - other.x
            y = self.y - other.y
            z = self.z - other.z
            return Vector(x, y, z)
        else:
            return NotImplemented

    def __mul__(self, other):
        if isinstance(other, (int, float)):
            x = self.x * other
            y = self.y * other
            z = self.z * other
            return Vector(x, y, z)
        else:
            return NotImplemented

    def __rmul__(self, other):
        if isinstance(other, (int, float)):
            x = self.x * other
            y = self.y * other
            z = self.z * other
            return Vector(x, y, z)
        else:
            return NotImplemented

    def __truediv__(self, other):
        if isinstance(other, (int, float)):
            if other == 0:
                return Vector(0, 0, 0)
            else:
                x = self.x / other
                y = self.y / other
                z = self.z / other
                return Vector(x, y, z)
        else:
            return NotImplemented

    def __rtruediv__(self, other):
        if isinstance(other, (int, float)):
            if self.x == 0:
                return Vector(0, 0, 0)
            else:
                x = other / self.x
                y = other / self.y
                z = other / self.z 
                return Vector(x, y, z)
        else:
            return NotImplemented  

    def __abs__(self):
        return (self.x**2 + self.y**2 + self.z**2)**0.5

    def normalise(self):
        return self / abs(self)